ml-agents/mlagents/trainers/stats.py

```diff
@@ -47,10 +59,43 @@ def write_text(self, category: str, text: str, step: int) -> None:
         self.summary_writers[category].add_summary(text, step)
 
 
-class StatsSummary(NamedTuple):
-    mean: float
-    std: float
-    num: int
+class CSVWriter(StatsWriter):
+    def __init__(self, base_dir: str):
+        # We need to keep track of the fields in the CSV, as all rows need the same fields.
+        self.csv_fields: Dict[str, List[str]] = {}
+        self.base_dir: str = base_dir
+
+    def write_stats(
+        self, category: str, values: Dict[str, StatsSummary], step: int
+    ) -> None:
+        self._maybe_create_csv_file(category, list(values.keys()))
+        row = [str(step)]
+        # Only record the stats that showed up in the first row.
+        for key in self.csv_fields[category]:
+            _val = values.get(key, None)
+            row.append(str(_val.mean) if _val else "None")
+        with open(self._get_filepath(category), "a") as file:
+            writer = csv.writer(file)
+            writer.writerow(row)
+
+    def _maybe_create_csv_file(self, category: str, keys: List[str]) -> None:
+        if category not in self.csv_fields:
+            summary_dir = self.base_dir
+            os.makedirs(summary_dir, exist_ok=True)
+            self.csv_fields[category] = keys
+            with open(self._get_filepath(category), "w") as file:
+                title_row = ["Steps"]
+                title_row.extend(keys)
+                writer = csv.writer(file)
+                writer.writerow(title_row)
+                self.csv_fields[category] = keys
```

Comment:
nit: redundant? You do the same before the `with...`

Reply:
Good catch  - removed